controllers/front/CartController.php

```diff
@@ -78,23 +78,91 @@ public function initContent()
             'static_token' => Tools::getToken(false),
         ]);
 
-        if (Tools::getValue('refreshajax')) {
-            ob_end_clean();
-            header('Content-Type: application/json');
-            die(json_encode([
-                'cart_detailed' => $this->render('checkout/_partials/cart-detailed.tpl'),
-                'cart_detailed_totals' => $this->render('checkout/_partials/cart-detailed-totals.tpl'),
-                'cart_summary_items_subtotal' => $this->render('checkout/_partials/cart-summary-items-subtotal.tpl'),
-                'cart_summary_totals' => $this->render('checkout/_partials/cart-summary-totals.tpl'),
-                'cart_voucher' => $this->render('checkout/_partials/cart-voucher.tpl'),
+        $this->setTemplate('checkout/cart.tpl');
+    }
+
+    public function displayAjaxUpdate()
+    {
+        $this->updateCart();
+
+        if (!$this->errors) {
+            $this->ajaxDie(Tools::jsonEncode([
+                'success' => true,
+                'id_product' => $this->id_product,
+                'id_product_attribute' => $this->id_product_attribute
+            ]));
+        } else {
+            $this->ajaxDie(Tools::jsonEncode([
+                'hasError' => true,
+                'errors' => [$this->l('Something went wrong during cart update')],
             ]));
         }
+    }
 
-        $this->setTemplate('checkout/cart.tpl');
+    public function displayAjaxRefresh()
+    {
+        ob_end_clean();
+        header('Content-Type: application/json');
+        $this->ajaxDie(Tools::jsonEncode([
+            'cart_detailed' => $this->render('checkout/_partials/cart-detailed.tpl'),
+            'cart_detailed_totals' => $this->render('checkout/_partials/cart-detailed-totals.tpl'),
+            'cart_summary_items_subtotal' => $this->render('checkout/_partials/cart-summary-items-subtotal.tpl'),
+            'cart_summary_totals' => $this->render('checkout/_partials/cart-summary-totals.tpl'),
+            'cart_voucher' => $this->render('checkout/_partials/cart-voucher.tpl'),
+        ]));
+    }
+
+    public function displayAjaxProductrefresh()
+    {
+        $url = $this->context->link->getProductLink(
+            $this->id_product,
+            null,
+            null,
+            null,
+            $this->context->language->id,
+            null,
+            (int)Product::getIdProductAttributesByIdAttributes($this->id_product, Tools::getValue('group')),
```

Comment:
It redirects to the default combination.
We have to display a warning if it's not available. Thanks for notice this.
I've created a [trello card](https://trello.com/c/BK6Ua0us)